internal/sync/lifecycle.go

```diff
@@ -55,85 +60,131 @@ const (
+	stopCh   chan struct{}
+	errored  atomic.Bool
 	startErr error
 	stopErr  error
 }
 
+// Once returns a lifecycle controller.
+// 0. The observable lifecycle state must only go forward from birth to death.
+// 1. Start() must block until the state is >= Running
+// 2. Stop() must block until the state is >= Stopped
+// 3. Stop() must pre-empt Start() if it occurs first
+// 4. Start() and Stop() may be backed by a do actual work function, and that
+//    function must be called at most once.
+func Once() LifecycleOnce {
+	return &lifecycleOnce{
+		startCh: make(chan struct{}, 0),
+		stopCh:  make(chan struct{}, 0),
+	}
+}
+
 // Start will run the `f` function once and return the error.
 // If Start is called multiple times it will return the error
 // from the first time it was called.
-func (l *LifecycleOnce) Start(f func() error) error {
-	l.lock.Lock()
-	defer l.lock.Unlock()
-
-	// If we've already moved on from the idle state we've either
-	// called the start function already, or called the stop function
-	// in which case we should exit now and return the result of the
-	// last start command (or nil).
-	if LifecycleState(l.state.Load()) != Idle {
+func (l *lifecycleOnce) Start(f func() error) error {
+	if l.starting.Swap(true) {
+		<-l.startCh
 		return l.startErr
 	}
-
-	// Set a nil function to an empty function
-	if f == nil {
-		f = func() error { return nil }
+	if f != nil {
+		l.startErr = f()
 	}
-
-	l.state.Store(int32(Starting))
-	l.startErr = f()
-	if l.startErr == nil {
-		l.state.Store(int32(Running))
-	} else {
-		l.state.Store(int32(Errored))
+	// skip forward to error state
+	if l.startErr != nil {
+		l.errored.Store(true)
+		l.stopped.Store(true)
+		l.stopping.Store(true)
+		close(l.stopCh)
 	}
+	l.started.Store(true)
+	close(l.startCh)
 
 	return l.startErr
 }
 
+func (l *lifecycleOnce) WhenRunning(ctx context.Context) error {
+	if !l.stopping.Load() && l.started.Load() {
+		return nil
+	}
+
+	select {
+	case <-l.startCh:
+		return nil
+	case <-l.stopCh:
+		return ErrAlreadyStopped
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
 // Stop will run the `f` function once and return the error.
 // If Stop is called multiple times it will return the error
 // from the first time it was called.
-func (l *LifecycleOnce) Stop(f func() error) error {
-	l.lock.Lock()
-	defer l.lock.Unlock()
-
-	// If the lifecycle is already stopped or errored, don't execute the stop func
-	if LifecycleState(l.state.Load()) == Stopped || LifecycleState(l.state.Load()) == Errored {
+func (l *lifecycleOnce) Stop(f func() error) error {
+	if l.stopping.Swap(true) {
+		<-l.stopCh
 		return l.stopErr
 	}
 
-	if f == nil {
-		f = func() error { return nil }
+	// Pre-empt start
+	if !l.starting.Swap(true) {
+		l.started.Store(true)
+		close(l.startCh)
 	}
 
-	l.state.Store(int32(Stopping))
-	l.stopErr = f()
-	if l.stopErr == nil {
-		l.state.Store(int32(Stopped))
-	} else {
-		l.state.Store(int32(Errored))
+	if l.started.Swap(true) {
```

Comment:
The test would pass without an assertion of "ExpectedFinishOrder" because we changed the state assertion from equal to greater than, so when the 'started' process finish it asserts that the state is greater than 'started', which it will be because the state will be 'stopped'